lib/experimental/results.nim

```diff
@@ -0,0 +1,187 @@
+##[
+This module implements a result-type which encapsulates both a success and a
+failure in one value. Only one of the two states can be active.
+
+Based on https://github.com/disruptek/badresults, which was itself based on
+https://github.com/arnetheduck/nim-result.
+
+Exception translation
+=====================
+
+Using `get` or `value` raises an exception if the `Result` stores a failure.
+For `Result[T, E]` the exception raised is:
+ * if `E` is a `ref` type that inherits from `Exception`, the `error` value
+ * if a custom overload of `toException <#toException, E>`_ exists, the result
+  of `toException(r)`
+ * otherwise, a new `ResultError[E]` with `error` as the payload
+
+]##
+
+type
+  ResultError*[E] = ref object of ValueError
```

Comment:
Should `ResultError` be a `object of ValueError` instead? With the current version, wanting to use the underlying `object` type requires some helper logic

Reply:
Yea, it should be converted to `object`.